src/main/java/com/google/api/generator/engine/ast/Reference.java

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.api.generator.engine.ast;
+
+import com.google.auto.value.AutoValue;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@AutoValue
+public abstract class Reference {
+  public abstract Class clazz();
+
+  public abstract List<Reference> generics();
+
+  public static Reference withClazz(Class clazz) {
+    return builder().setClazz(clazz).build();
+  }
+
+  public static Builder builder() {
+    return new AutoValue_Reference.Builder().setGenerics(Collections.emptyList());
+  }
+
+  public String name() {
+    StringBuilder sb = new StringBuilder();
+    sb.append(clazz().getSimpleName());
+    if (!generics().isEmpty()) {
+      sb.append("<");
+      for (int i = 0; i < generics().size(); i++) {
+        Reference r = generics().get(i);
+        sb.append(r.name());
+        if (i < generics().size() - 1) {
+          sb.append(", ");
+        }
+      }
+      sb.append(">");
+    }
+    return sb.toString();
+  }
+
+  public Set<String> fullNames() {
+    Set<String> allNames =
+        Stream.<String>of(clazz().getCanonicalName()).collect(Collectors.toSet());
```

Comment:
For getting allNames, if I'm understanding correctly, can we do this?
```
Set<String> allNames = new HashSet<>(Arrays.asList(clazz().getCanonicalName())); 
```

Reply:
That is also an option. It turns out that this method is not needed, so I've removed `fullNames()`.

Reply:
👍